User command-clicks on fragment URLS should not navigate on current page.

megangardner opened the pull request:
#### b872b71975a2c60bb1496dbddf05dcaf987dc525
<pre>
User command-clicks on fragment URLS should not navigate on current page.
<a href="https://bugs.webkit.org/show_bug.cgi?id=267168">https://bugs.webkit.org/show_bug.cgi?id=267168</a>
<a href="https://rdar.apple.com/119079650">rdar://119079650</a>

Reviewed by Wenson Hsieh and Alex Christensen.

In <a href="https://commits.webkit.org/270416@main">https://commits.webkit.org/270416@main</a> a change was made for performance
to asynchronously  DecidePolicyForNavigationAction for fragment loading. While this is
a good speedup, it breaks the long-standing user interaction of command-clicking on
a fragment link to load that link in a background tab, while staying at the current
scroll location on the current page. This is a client behavior, and with this change, the
client no longer had control to keep the navigation to the fragment from happening.
We do not want to lose this speedup, so only revert to the previously synchronous behavior
when the fragment URL is being loaded from a user gesture.

* LayoutTests/fast/dom/no-scroll-when-command-click-fragment-URL-expected.txt: Added.
* LayoutTests/fast/dom/no-scroll-when-command-click-fragment-URL.html: Added.
* Source/WebKit/WebProcess/WebCoreSupport/WebFrameLoaderClient.cpp:
(WebKit::WebFrameLoaderClient::dispatchDecidePolicyForNavigationAction):

Canonical link: <a href="https://commits.webkit.org/272906@main">https://commits.webkit.org/272906@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b69fc4e25118dcfe0743eb76a5e55b4e5b9b6ff6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/33467 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/12238 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/35385 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/36091 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/30401 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/14683 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/9386 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/29525 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/33942 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/10375 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/29866 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/9017 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/9168 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/29857 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/37415 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/30442 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/30206 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/35264 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/9259 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/7200 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/33139 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/11014 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/29542 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7761 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/9844 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/9991 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on LayoutTests/fast/dom/no-scroll-when-command-click-fragment-URL.html:
```diff
@@ -0,0 +1,60 @@
+
+    var scrollOffset = window.pageYOffset;
+
+    UIHelper.activateElement(document.getElementById("LinkToClick"), ["metaKey"]); // metaKey = commandKey
```
Missing an `await` here.
```suggestion
    await UIHelper.activateElement(document.getElementById("LinkToClick"), ["metaKey"]); // metaKey = commandKey
```

whsieh commented on LayoutTests/fast/dom/no-scroll-when-command-click-fragment-URL.html:
```diff
@@ -0,0 +1,60 @@
+
+  window.addEventListener('load', () => {
+    runTest();
+  }, false);
```
Would be simpler as:

```suggestion
addEventListener('load', runTest);
```

whsieh commented on LayoutTests/fast/dom/no-scroll-when-command-click-fragment-URL.html:
```diff
@@ -0,0 +1,60 @@
+
+    </div>
+</body>
+</html>
```
There's no corresponding `<html>` tag for this closing tag.

whsieh commented on LayoutTests/fast/dom/no-scroll-when-command-click-fragment-URL.html:
```diff
@@ -0,0 +1,60 @@
+    testRunner.waitUntilDone();
+}
+
+  
```
Nit - extra newline

whsieh commented on LayoutTests/fast/dom/no-scroll-when-command-click-fragment-URL.html:
```diff
@@ -0,0 +1,60 @@
+ async function runTest()
+ {
+    if (!testRunner.runUIScript)
+        return;
```
Nit - the indentation is a bit strange in this file (there's a mix of 2, 3, and 4-space indents)

whsieh commented on Source/WebKit/WebProcess/WebCoreSupport/WebFrameLoaderClient.cpp:
```diff
@@ -161,7 +161,7 @@ void WebFrameLoaderClient::dispatchDecidePolicyForNavigationAction(const Navigat
     if (policyDecisionMode == PolicyDecisionMode::Synchronous) {
 #if PLATFORM(COCOA)
-        if (!linkedOnOrAfterSDKWithBehavior(SDKAlignedBehavior::AsyncFragmentNavigationPolicyDecision)) {
+        if (navigationAction.userGestureToken() || !linkedOnOrAfterSDKWithBehavior(SDKAlignedBehavior::AsyncFragmentNavigationPolicyDecision)) {
```
I wonder if this should check for `processingUserGesture()` instead of  `userGestureToken()`...

The former is a bit more restrictive, and (seems to) take gestures that only trigger media playback into account.

hortont424 commented on LayoutTests/fast/dom/no-scroll-when-command-click-fragment-URL.html:
```diff
@@ -0,0 +1,60 @@
+<!DOCTYPE html>
+<meta charset="utf-8" />
+<title>Fragment Navigation - ensure opening a fragment URL does not navigate on activating page.</title>
```
This title sounds wrong

megangardner replied:
What part is confusing? We need to make sure that when you command-click on a link that it doesn't actually navigate to the fragment on the page that you are currently on.

hortont424 replied:
(I explained offline that I was confused by "activating page")

hortont424 commented on LayoutTests/fast/dom/no-scroll-when-command-click-fragment-URL.html:
```diff
@@ -0,0 +1,60 @@
+
+    await UIHelper.ensurePresentationUpdate();
+
+    if (window.pageYOffset != scrollOffset) {
```
lots of extraneous curlies

hortont424 commented on LayoutTests/fast/dom/no-scroll-when-command-click-fragment-URL.html:
```diff
@@ -0,0 +1,60 @@
+</head>
+
+<body>
+    <a href="#fragment" id = "LinkToClick">Link to Fragment</a>
```
no spaces around =

hortont424 commented on LayoutTests/http/tests/navigation/fragment-navigation-policy-ignore.html:
```diff
@@ -0,0 +1,25 @@
+
+if (window.testRunner) {
+    testRunner.setCustomPolicyDelegate(true, false);
+    if (testRunner.skipPolicyDelegateNotifyDone) { testRunner.skipPolicyDelegateNotifyDone() }
```
no curlies and not on the same line IMO

hortont424 commented on LayoutTests/fast/dom/no-scroll-when-command-click-fragment-URL.html:
```diff
@@ -0,0 +1,60 @@
+<body>
+    <a href="#fragment" id = "LinkToClick">Link to Fragment</a>
+    <div id="target" style="padding-top: 10000px;">
+    
```
why the blank lines?

megangardner replied:
readbility

hortont424 reviewed (commented):
Some grammar errors in "it breaks the long-standing user interaction of command-clicking on
a fragment link to load that link in a background tab, while staying are the current
scroll location on the current page", but also I think it should be more clear that it break's the *client's* ability to do these things (these aren't WebKit behaviors, they're behaviors of some WebKit clients, and we've taken away their ability to implement them).

megangardner committed:
User command-clicks on fragment URLS should not navigate on current page.
https://bugs.webkit.org/show_bug.cgi?id=267168
rdar://119079650

Reviewed by Wenson Hsieh and Alex Christensen.

In https://commits.webkit.org/270416@main a change was made for performance
to asynchronously  DecidePolicyForNavigationAction for fragment loading. While this is
a good speedup, it breaks the long-standing user interaction of command-clicking on
a fragment link to load that link in a background tab, while staying at the current
scroll location on the current page. This is a client behavior, and with this change, the
client no longer had control to keep the navigation to the fragment from happening.
We do not want to lose this speedup, so only revert to the previously synchronous behavior
when the fragment URL is being loaded from a user gesture.

* LayoutTests/fast/dom/no-scroll-when-command-click-fragment-URL-expected.txt: Added.
* LayoutTests/fast/dom/no-scroll-when-command-click-fragment-URL.html: Added.
* Source/WebKit/WebProcess/WebCoreSupport/WebFrameLoaderClient.cpp:
(WebKit::WebFrameLoaderClient::dispatchDecidePolicyForNavigationAction):

Canonical link: https://commits.webkit.org/272906@main
